torch/csrc/jit/codegen/cuda/predicate_compute.cpp

```diff
@@ -308,6 +308,105 @@ UnswitchPredicate::UnswitchPredicate(
   openLoop(unrolled_loop);
 }
 
+kir::Bool* VectorizePredicate::get(
+    const std::vector<kir::ForLoop*>& outer_loops,
+    kir::ForLoop* vectorized_loop,
+    const IterDomainMap& p2c_root_map,
+    const ComputeAtRootDomainMap& ca_root_map) {
+  FUSER_PERF_SCOPE("VectorizePredicate::get");
+
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+  VectorizePredicate up(
+      outer_loops, vectorized_loop, p2c_root_map, ca_root_map);
+  return up.vectorize_pred_;
+}
+
+void VectorizePredicate::predicateOn(kir::Expr* tv_expr) {
+  FUSER_PERF_SCOPE("UnswitchPredicate::predicateOn");
+
+  if (for_loops_.empty()) {
+    return;
+  }
+
+  if (!tv_expr->isA<kir::UnaryOp>() ||
+      tv_expr->as<kir::UnaryOp>()->operation() != UnaryOpType::Set) {
+    return;
+  }
+
+  auto unaryOp = tv_expr->as<kir::UnaryOp>();
+  auto in_tv = unaryOp->in()->as<kir::TensorView>();
+  auto out_tv = unaryOp->out()->as<kir::TensorView>();
+
+  auto pred_contiguity = out_tv->domain()->contiguity();
+  for (auto inp : tv_expr->inputs()) {
+    if (auto inp_tv = dynamic_cast<kir::TensorView*>(inp)) {
+      if (inp_tv->domain()->hasRFactor() ||
+          inp_tv->memoryType() == MemoryType::Shared ||
+          inp_tv->memoryType() == MemoryType::Local) {
+        continue;
+      } else {
+        pred_contiguity = IndexCompute::contiguityAnd(
+            pred_contiguity,
+            IndexCompute::contiguityPasC(inp_tv->domain(), out_tv->domain()));
+      }
+    }
+  }
+
+  bool isVectorizedRead = out_tv->memoryType() == MemoryType::Local &&
+      in_tv->memoryType() == MemoryType::Global;
+
+  auto pred_inds = (isVectorizedRead)
+      ? Index::getProducerRootVectIndices(
+            in_tv, out_tv, for_loops_, pred_contiguity, ca_root_map_)
+      : Index::getConsumerRootVectIndices(
+            out_tv, for_loops_, pred_contiguity, ca_root_map_);
+
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  kir::Val* index = nullptr;
+  for (auto idx : pred_inds) {
+    index = (index == nullptr) ? idx : ir_builder.addExpr(index, idx);
+  }
+
+  //! check base_index % vector_size == 0
```

Comment:
Should this be done at runtime not in the kernel? Do you have a path to take if this isn't true in the kernel?